klippy/extras/heaters.py

```diff
@@ -5,16 +5,107 @@
+        time_diff = read_time - self.last_temp_time
+        next_smoothed_temp = self._smooth(time_diff, temp, self.smoothed_temp)
+        # Temperature Derivatives
+        self.temp_slope = self._derivative(self.last_temp, temp,
+                        time_diff, self.temp_slope)
+        smoothed_derivative = self._derivative(self.smoothed_temp,
+                        next_smoothed_temp, time_diff, self.smoothed_temp_slope)
+        self.smoothed_temp_slope = self._smooth(time_diff, smoothed_derivative,
+                         self.smoothed_temp_slope)
+        # Store state for next measurement
+        self.last_temp = temp
+        self.smoothed_temp = next_smoothed_temp
+        self.last_temp_time = read_time
+        self.time_diff = time_diff
+        # Call any installed callback
+        if self._callback:
+            self._callback(read_time, temp)
+    def _smooth(self, time_diff, value, smoothed_value):
+        # Time based smoothing of a value
+        adj_time = min(time_diff * self.inv_smooth_time, 1.)
+        value_diff_smoothed = value - smoothed_value
+        return smoothed_value + (value_diff_smoothed * adj_time)
+    def _derivative(self, prev_temp, next_temp, time_diff, previous_deriv):
+        # Temperature Derivative
+        temp_diff = next_temp - prev_temp
+        if time_diff >= self.min_deriv_time:
+            return temp_diff / time_diff
+        else:
+            return (previous_deriv * (self.min_deriv_time - time_diff)
+                        + temp_diff) / self.min_deriv_time
+    def setup_callback(self, callback):
+        self._callback = callback
+    # return the hardware sensor object
+    def get_sensor(self):
+        return self.sensor
+    # returns the smoothed temperature
+    def get_smoothed_temp(self):
+        return self.smoothed_temp
+    def get_smooth_time(self):
+        return self.smooth_time
+    # returns the last raw temperature reading without smoothing
+    def get_last_temp(self):
+        return self.last_temp
+    # time when the last temperature reading was taken
+    def get_last_temp_time(self):
+        return self.last_temp_time
+    # time delta between the last 2 temperature readings
+    def get_last_temp_time_delta(self):
+        return self.time_diff
+    # returns the rate of change of the raw temperature in degrees C/s
+    def get_temp_slope(self):
+        return self.temp_slope
+    # returns the rate of change of the smoothed temperature in degrees C/s
+    def get_smoothed_temp_slope(self):
+        return self.smoothed_temp_slope
+    # return the configured min temp
+    def get_status(self, eventtime):
+        return {
+            'temperature': self.smoothed_temp,
+            'last_temperature': self.last_temperature,
+            'temperature_slope': self.temp_slope,
+        }
+
+######################################################################
+# Heater
+######################################################################
+
```

Comment:
extra blank line should be removed

Reply:
Actually, the style seems to be to have a blank line. The merge just looks more complicated than it should be.